custom/conf/app.example.ini

```
@@ -1244,6 +1244,9 @@ LEVEL = Info
 ;; Change the sort type of the explore pages.
 ;; Default is "recentupdate", but you also have "alphabetically", "reverselastlogin", "newest", "oldest".
 ;EXPLORE_PAGING_DEFAULT_SORT = recentupdate
+;;
+;; Whether to prefer all timestamps to be rendered as absolute. Setting this to false means some timestamps would render as relative.
+;PREFER_ABSOLUTE_TIMESTAMPS = false
```

On a related note, wouldn't it be better to name the setting
```suggestion
;TIMESTAMP_FORMAT = relative
```
with `relative,absolute` being the possible options (for now)

---

I don't think other options exist

---

I can think of a few at the top of my head: `relative_v2, absolute_v2, mixed, …`

---

If relative is set here, some timestamps still render as absolute though (e.g. joined on date in profile page)

---

Yeah, but those are either intended or a bug.
Either way, to me it makes sense despite some minor exceptions.

---

We can even think about whether a relative timestamp wouldn't make sense in your example as well:
What is your information need when looking at the `joined at`?
The exact date the user joined, or how long they already exist?

---

There are many places we use absolute dates. I don't know why some are absolute and some are relative.

`TIMESTAMP_FORMAT` seems like it applies to all timestamps whereas having the word `PREFER` suggests it's a best-effort kind of thing. How about `PREFERRED_TIMESTAMP_FORMAT`? (though I still think a boolean fits better here)

---

I could also agree a string config option might be more flexible.

For me, I guess I would like to use `TIMESTAMP_FORMAT = sql` to make the UI always show `2001-02-03 04:05:06` if it is possible. (well, just my personal preference, no need to treat it seriously)

----

Actually, maybe we need a document and a test page (eg: `devtest`) to describe and show (even test) the "relative-time" component behaviors and the effect of the config options.

---

@delvh @wxiaoguang will you block it as-is? I'd much rather not change it due to reasons I mentioned before

---

I don't really care about the name, what's more important is that it is a string based attribute instead of a boolean one.

---

I just don't see why any other string would ever be viable. A timestamp is either absolute or relative. wxiaoguang's example is more about the rendered format which is different than the tense (in the example there's an absolute tense).

---

No block from my side, I am neutral for it (just expressing my thoughts when I saw the discussion)

---

Yes, but exactly this "format" is why I think it won't hurt to have a string based one.
You cannot know what formats will be added in the future, I only know that as you've seen yourself with wxiaoguang's example, alternative formats are possible.

---

We have to differentiate between format and tense. There can be many different formats for each tense but there are only two tenses possible, either absolute or relative.

---

Do we?
Why aren't these two the same thing?

---

For example absolute dates:
- 2023-12-31
- December 31st, 2023
- Dec 31, 2023
- 31/12/2023
- 12/31/2023
- ...

All of the above examples are of different formats but the same tense.

In the relative tense, we usually have 1 format per language. In the end this setting is about the tense, not the format.

---

That wasn't my question.
My question is what's the benefit in separating the two?

Here's what I mean:
What do you want to achieve when you configure this setting?

Well, you want to influence how all (or at least most) timestamps are displayed.
In other words, the `format` you keep mentioning.
**The tense can be completely ignored.**
It is simply a coincidence that the format you choose is absolute or relative.
In the end, all that matters is that you choose the one you like the most.

So, what's the point of a `tense` setting if we only need the `format` setting?

---

To control the format, much more logic is needed. This setting's intent is only to require absolute dates, so as to not require hovering to see the date https://github.com/go-gitea/gitea/issues/22493.

---

But is that a job of this PR?
As I said, I don't see a reason not to use a string based one, then we can still add the additional logic later on.

---

The job of this PR is to switch from relative tense to absolute tense, not to control the format. The format would have to be a different setting.

---

https://github.com/go-gitea/gitea/pull/28657#discussion_r1438856953

---

It simply doesn't make sense for me to consider the two to be separate entities.
**There's no need for a second setting.**

---

The original issue requests a change in tense, not format. I want the format to remain client-rendered and localized by the browser such that the server only controls the tense as it does today.

---

I have nothing against that, I don't see a reason why this shouldn't be possible with a string based option too.
The only thing that would change is that instead of `if (preferAbsoluteTimestamp)`, you'd check for `if (preferredTimestampFormat == "absolute")`
Nevertheless, what's stopping you from converting the boolean option to a string based one?

---

Because absolute and relative are not formats. This is misleading.

---

It is a format in the sense that these are the formats we offer by default.

---

How about `PREFERRED_TIMESTAMP_TENSE`?

---

Sounds good

---

Do you agree it should be `mixed | absolute`? `relative` alone doesn't exist today

---

Okay, if you add a comment why `relative` does not yet exist.

---

My current description is:

- `PREFERRED_TIMESTAMP_TENSE`: **mixed**: The tense all timestamps should be rendered as. Possible values are `absolute` time (i.e. 1970-01-01, 11:59) or `mixed` which means some timestamps would render as relative (i.e. 2 days ago).

---

Today some of our timestamps are absolute and some are relative. I don't know why.

---

~~some~~ most

---

Done https://github.com/go-gitea/gitea/pull/28657/commits/0504a8c68520e7c8703091cbf9758ae42f1398ac